### DISCUSSION TITLE
Run workflow only if secret is present

### DISCUSSION THREAD

**Discussion by Zethson** [community/community#25444] [Actions]
<p>Dear everyone,</p>
<p>I want a workflow to only be run if a secret is available.<br>
If not, I want it to either just exit with success or do something else.</p>
<p>In pseudocode:</p>
<p>if secret.available:<br>
run A<br>
else:<br>
run B</p>
<p>Does anybody have any ideas and examples?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/zethson">@zethson</a>,</p>
<p>You can use the APIs “<a href="https://docs.github.com/en/rest/reference/actions#list-repository-secrets" rel="nofollow noopener"><strong>List repository secrets</strong></a>” and “<a href="https://docs.github.com/en/rest/reference/actions#list-organization-secrets" rel="nofollow noopener"><strong>List organization secrets</strong></a>” to get the total count of the repository-level secrets and organization-level secrets. Then according to the total count to determine whether to continue the subsequent jobs.</p>
<p>To run these two APIs, you need to create a personal access token (PAT) that has the <strong>repo</strong> scope (access repository-level secrets) and <strong>admin:org</strong> scope (access organization-level secrets),  and set this PAT as a secret in the repository or in the organization. Then you can use this PAT to authenticate in the workflow.</p>
<p>Due to the added authorization PAT, there is an existing secret at least. If you want to ignore the PAT in the secrets, you can set the available count to be more than 1. If the count is equal or less than 1, you can determine it as “No available secrets”.<br>
Here is an example as reference:<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/TestWorkflowsRML/RunAPI/actions/runs/208023621/workflow" target="_blank" rel="nofollow noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/46307996/181091155-1ba9f348-3294-4d71-b15f-87f79d5cb711.png" class="thumbnail onebox-avatar" width="420" height="420">

<h3><a href="https://github.com/TestWorkflowsRML/RunAPI/actions/runs/208023621/workflow" target="_blank" rel="nofollow noopener">TestWorkflowsRML/RunAPI</a></h3>

<p>Contribute to TestWorkflowsRML/RunAPI development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<pre><code class="lang-yaml">jobs:
  job1:
    runs-on: ubuntu-latest
    steps:
      - name: Count of secrets
        run: |
          echo "➖➖➖➖➖➖➖➖ Count of repository secrets ➖➖➖➖➖➖➖➖"
          repo_response=$(curl \
          -H "Authorization: token ${{ secrets.MY_GITHUB_PAT }}" \
          -H "Accept: application/vnd.github.v3+json" \
          https://api.github.com/repos/${{ github.repository }}/actions/secrets)
          
          echo "$repo_response"
          
          repo_count=$(echo "$repo_response" | jq '.total_count')
          echo "The count of secrets in current repository is $repo_count."
          
          echo "➖➖➖➖➖➖➖➖ Count of organization secrets ➖➖➖➖➖➖➖➖"
          org_response=$(curl \
          -H "Authorization: token ${{ secrets.MY_GITHUB_PAT }}" \
          -H "Accept: application/vnd.github.v3+json" \
          https://api.github.com/orgs/${{ github.repository_owner }}/actions/secrets)
          
          echo "$org_response"
          
          org_count=$(echo "$org_response" | jq '.total_count')
          echo "The count of secrets in current organization is $org_count."
          
          echo "➖➖➖➖➖➖➖➖ Count of total secrets ➖➖➖➖➖➖➖➖"
          total_count=$(( $repo_count + $org_count ))
          echo "The total count of secrets in is $total_count."
          
          if [[ $total_count -le 1 ]]; then
            echo "::error::No available secrets!"
            exit 1
          fi
  job2:
    needs: [job1]
    runs-on: ubuntu-latest
    steps:
      - name: run this job
        run: echo "This job will run if there are available secrets."

  job3:
    needs: [job1]
    runs-on: ubuntu-latest
    steps:
      - name: run this job
        run: echo "This job will run if there are available secrets."
</code></pre>
<p>When the total count of the secrets (repository-level and organization-level) is equal or less than 1, set <strong>job1</strong> to be failed with the error message “<strong>No available secrets!</strong>”. Then the subsequent jobs (<strong>job2</strong> and <strong>job3</strong>) will be skipped.</p>

**Comment by GabLeRoux** [comment]
<p>Here’s a related question with alternative solutions: <a href="https://github.community/t/how-can-i-test-if-secrets-are-available-in-an-action/17911" class="inline-onebox">How can I test if secrets are available in an action?</a></p>
